src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs

```diff
@@ -0,0 +1,776 @@
+        )?;
+        info!(
+            "{}: waiting for incoming connection from the slave...",
+            NET_DRIVER_NAME
+        );
+        let (master, avail_features) = listener.accept()?;
+        info!("{}: connection to slave is ready.", NET_DRIVER_NAME);
+        Self::new(
+            master,
+            avail_features,
+            listener,
+            guest_mac,
+            queue_sizes,
+            epoll_mgr,
+        )
+    }
+
+    fn activate_slave<AS, Q, R>(
+        &mut self,
+        handler: &VhostUserNetHandler<AS, Q, R>,
+    ) -> ActivateResult
+    where
+        AS: DbsGuestAddressSpace,
+        Q: QueueT + Send + 'static,
+        R: GuestMemoryRegion + Sync + Send + 'static,
+    {
+        trace!(target: "vhost-net", "{}: VhostUserNetDevice::activate_slave()", self.id);
+        // Safe because the `intr_group` has been initialized in
+        // VirtioDevice::activate().
+        let mut config = EndpointParam {
+            virtio_config: &handler.config,
+            intr_evts: handler.config.get_queue_interrupt_eventfds(),
+            queue_sizes: &self.device_info.queue_sizes,
+            features: self.get_acked_features(),
+            protocol_flag: 0,
+            dev_protocol_features: Self::get_dev_protocol_features(),
+            reconnect: false,
+            backend: None,
+            init_queues: self.curr_queues,
+            slave_req_fd: None,
+        };
+        config.set_protocol_mq();
+        // Do negotiate with the vhost-user slave
+        loop {
+            match self.endpoint.negotiate(&config, None) {
+                Ok(_) => break,
+                Err(VirtioError::VhostError(VhostError::VhostUserProtocol(err))) => {
+                    if err.should_reconnect() {
+                        // Fall through to rebuild the connection.
+                        warn!(
+                            "{}: socket disconnected while initializing the connnection, {}",
+                            self.id, err
+                        );
+                    } else {
+                        error!("{}: failed to setup connection, {}", self.id, err);
+                        return Err(VhostError::VhostUserProtocol(err).into());
+                    }
+                }
+                Err(err) => {
+                    error!("{}: failed to setup connection, {}", self.id, err);
+                    return Err(err.into());
+                }
+            }
+            // Do reconnect
+            // Wait 100ms for the next connection
+            let delay = Duration::from_millis(100);
+            std::thread::sleep(delay);
+            // The underlying communication channel has been disconnected,
+            // recreate it again.
+            let (master, avail_features) = self.listener.accept()?;
+            if !avail_features & self.device_info.acked_features() != 0 {
+                error!("{}: Virtio features changed when reconnecting.", self.id);
```

Comment:
Would be helpful to know which features mismatched.

Reply:
Fixed.